### PR TITLE
[PLATFORM-904]: Add support for middleware

### PR DESCRIPTION
**Pull request by oliverbrowneprima**
https://prima-assicurazioni-spa.myjetbrains.com/youtrack/issue/PLATFORM-904

### CODE REVIEW

File: src/errors.rs [hunk 1]
```diff
@@ -35,6 +35,42 @@ pub enum PrimaBridgeError {
     Utf8Error { source: Utf8Error },
 }
 
+impl From<reqwest::Error> for PrimaBridgeError {
+    fn from(error: reqwest::Error) -> Self {
+        Self::HttpError {
+            url: error.url().cloned().unwrap_or_else(default_url),
```
Comment (cottinisimone) [inline src/errors.rs:41]: I don't like so much the `default_url`. I think this should be implemented using `TryFrom`

File: src/errors.rs [hunk 2]
```diff
@@ -35,6 +35,42 @@ pub enum PrimaBridgeError {
...
+            source: error,
+        }
+    }
+}
+
+pub enum PrimaBridgeWithMiddlewareError {
```
Comment (cottinisimone) [inline src/errors.rs:47]: Why do we need this error? Why cannot `MiddlewareError(reqwest_middleware::Error),` stay in the `PrimaBridgeError`?

File: src/lib.rs [hunk 3]
```diff
@@ -41,15 +44,191 @@ mod response;
...
+    inner_client: T,
     endpoint: Url,
     #[cfg(feature = "auth0")]
     auth0_opt: Option<auth0::Auth0>,
 }
 
+/// A trait that abstracts the client used by the [BridgeImpl], such that both reqwest clients and reqwest
+/// clients with middleware can be used, more or less interchangeably.
+#[doc(hidden)]
+pub trait BridgeClient: Sealed {
```
Comment (cottinisimone) [inline src/lib.rs:63]: Superlike the `Sealed` approach

File: src/lib.rs [hunk 4]
```diff
@@ -63,3 +242,16 @@ impl Bridge {
         self.auth0_opt.as_ref().map(|auth0| auth0.token())
     }
 }
+
+mod sealed {
+    use crate::BridgeClient;
+
+    pub trait Sealed {}
```
Comment (cottinisimone) [inline src/lib.rs:249]: What about [this](https://crates.io/crates/sealed)?

**Review by cottinisimone — COMMENTED** [review]
Just few minor stuff. Great job :)

File: src/lib.rs (hunk 4 above)
Comment (cottinisimone) [reply src/lib.rs:249]: Just for the sake of being supersafe, have you checked that from outside you can used the types that implement the `Sealed` trait. I guess you can but a try could be worth it..

Comment (oliverbrowneprima) [reply src/lib.rs:249]: Do the existing tests cover the usage you are thinking? They're unable to implement `Sealed` but make use of sealed types, I think?

Comment (cottinisimone) [reply src/lib.rs:249]: Yes, you are right. Since tests under `tests` folder are able to access only the high level api i will say it's ok :)

**Review by cottinisimone — APPROVED** [review]

**Pull request merged**